src/yaml.cpp

```diff
@@ -104,30 +103,107 @@ estd::result<YAML::Node, estd::error> readYamlFile(std::string const & path) {
 	buffer << file.rdbuf();
 	return YAML::Load(buffer.str());
 }
-
-YamlResult<void> mergeYamlNodes(YAML::Node & map_a, YAML::Node map_b) {
-	// Check if the arguments are maps.
-	if (!map_a.IsMap() && !map_a.IsNull()) {
-		return YamlError{"tried to merge into a YAML node that is not a map"};
-	}
-	if (!map_b.IsMap() && !map_b.IsNull()) {
-		return YamlError{"tried to merge from a YAML node that is not a map"};
+// Merge Yaml Nodes of type Ordered Dictionary, i.e., sequence on Maps.
+YamlResult<void> mergeYamlOrderedDict(YAML::Node & map_a, YAML::Node map_b);
+
+inline YamlResult<void> mergeYamlOrderedDict(YAML::Node && map_a, YAML::Node map_b) {
+ 	return mergeYamlOrderedDict(map_a, map_b);
+ }
+
+// Merge Yaml Nodes of type Map , i.e., a key-value pair.
+YamlResult<void> mergeYamlMaps(YAML::Node & map_a, YAML::Node map_b);
+
+inline YamlResult<void> mergeYamlMaps(YAML::Node && map_a, YAML::Node map_b) {
+ 	return mergeYamlMaps(map_a, map_b);
+  }
+
+
+// Merge Yaml Nodes (Type: Map).
+YamlResult<void> mergeYamlMaps(YAML::Node & map_a, YAML::Node map_b) {
+	if (map_b.IsMap()){
+		for (YAML::const_iterator iterator = map_b.begin(); iterator != map_b.end(); iterator++) {
+			std::string key = iterator->first.as<std::string>();
+			YAML::Node value = iterator->second;
+			if (map_a[key] && map_a[key].IsMap()) {
+				auto merged = mergeYamlMaps(map_a[key], value);
+				if (!merged) {
+					merged.error().appendTrace({key, "", YAML::NodeType::Map});
+					return merged;
+				}
+			}
+			else if (value.Tag() == "!ordered_dict"  && map_a[key].Tag() == "!ordered_dict" && map_a[key]) {
+				auto merged = mergeYamlOrderedDict(map_a[key], value);
+				if (!merged) {
+					merged.error().appendTrace({key, "", YAML::NodeType::Map});
+					return merged;
+				}
+			}
+			else {
+			map_a[key] = value;
+			}
+		}
 	}
+	return estd::in_place_valid;
+}
 
-	for (YAML::const_iterator iterator = map_b.begin(); iterator != map_b.end(); iterator++) {
-		std::string key = iterator->first.as<std::string>();
-		YAML::Node value = iterator->second;
-		if (value.IsMap() && map_a[key]) {
-			auto merged = mergeYamlNodes(map_a[key], value);
-			if (!merged) {
-				merged.error().appendTrace({key, "", YAML::NodeType::Map});
-				return merged;
+// Merge Yaml Nodes (Type: Ordered Dictionary).
+YamlResult<void> mergeYamlOrderedDict(YAML::Node & map_a, YAML::Node map_b) {
+	std::size_t i = 0;
+	std::size_t j = 0;
+	std::size_t a_pointer = 0;
+	std::size_t b_pointer = 0;
```

Comment:
It is very hard to follow what is happening with the indices now.

I would propose to add some utility functions to make the code easier to reason about. In particular, I think it would help to have at least these helper functions:
* `std::optional<std::size_t> orderedDictFind(YAML::Node parent, std::string const & name)` 
* `void orderedDictUpdateOrInsert(YAML::Node parent, std::string const & name, YAML::Node contents)`

I think those functions will be much easier to verify on their own, and then the merge algorithm should become a lot easier to read too.